### PR TITLE
Additions to own type

### PR DESCRIPTION
**Pull request by gomoripeti**
With using our own abstract_expr() type I was hoping to get rid of some self-gradualizing errors, but there must be other reasons (eg see known problem added)

### CODE REVIEW

File: src/gradualizer_bin.erl [hunk 1]
```diff
@@ -7,7 +7,7 @@
 %% of the elements. The returned type is a normalized bitstring type.
 -spec compute_type(ExprOrPat) -> gradualizer_type:abstract_type()
         when ExprOrPat :: {bin, _, _},
-             ExprOrPat :: erl_parse:abstract_expr().
+             ExprOrPat :: gradualizer_type:abstract_expr() | gradualizer_type:af_pattern().
```
Comment (gomoripeti) [inline src/gradualizer_bin.erl]: I'm not sure explicitly added `af_pattern()` is necessary.

File: src/typechecker.erl [hunk 2]
```diff
@@ -85,7 +85,7 @@ compatible(Ty1, Ty2, TEnv) ->
 
 %% The first argument is a "compatible subtype" of the second.
 
--spec subtype(type() | gradualizer_type:af_character(), type(), #tenv{}) -> {true, any()} | false.
```
Comment (gomoripeti) [inline src/typechecker.erl]: `af_character()` is now fully part of `type()`

File: src/gradualizer_type.erl [hunk 3]
```diff
@@ -21,7 +21,7 @@
 -export_type([abstract_type/0]).
 
 %% Export the additional types that gradualizer uses
--export_type([af_character/0]).
+-export_type([af_pattern/0]).
```
Comment (zuiderkwast) [inline src/gradualizer_type.erl]: How about naming the exported type `abstract_pattern`, similarly to the other exported types?

**Comment by zuiderkwast** [comment]
Looks good so far. Is there more to come? bounded_fun, intersection?

File: src/gradualizer_bin.erl (hunk 1 above)
Comment (zuiderkwast) [reply src/gradualizer_bin.erl]: I though pattern is a subtype of expression, syntactically.

File: test/known_problems/should_pass/pattern_ann_type.erl [hunk 4]
```diff
@@ -0,0 +1,13 @@
+-module(pattern_ann_type).
```
Comment (josefs) [inline test/known_problems/should_pass/pattern_ann_type.erl]: I suggest you put this file addition in a separate patch.

Comment (gomoripeti) [reply test/known_problems/should_pass/pattern_ann_type.erl]: ok

(It is a simplification of the below error that comes out during self-gradualization, I thought that that error is related to the changes in this PR, but it turns out it is not
```
The pattern {var,_,'_'} on line {3345,14} doesn't have the type Pat :: gradualizer_type:af_pattern()
```
)

File: src/gradualizer_bin.erl (hunk 1 above)
Comment (gomoripeti) [reply src/gradualizer_bin.erl]: semantically they are different, that's why I felt the difference, but indeed syntactically they are not.
I even asked gradualizer... so I will remove this modification
```
1> typechecker:subtype(typelib:parse_type("gradualizer_type:af_pattern()"),typelib:parse_type("gradualizer_type:abstract_expr()"), {tenv, #{}, #{}}).
{true, ...}
```

**Comment by gomoripeti** [comment]
@zuiderkwast exactly bounded_fun, intersection will come too but I'm not sure if/when. so not in this PR.
